### PR TITLE
auto import all project for each workspace

### PR DESCRIPTION
**Pull request by Eskibear**
1. remove `import project`, `import All`, `remove project` commands
2. move `ProjectItem` to model folder, add other specific TreeItem type.
3. auto import projects for each workspace .

image

### CODE REVIEW

File: src/Utils.ts [hunk 1]
```diff
@@ -6,19 +6,19 @@ import * as os from "os";
...
+import { IArchetype, IArchetypeCatalogRoot, IArchetypes, IPomRoot } from "./model/XmlSchema";
 
 const EXTENSION_ID: string = "vscode-maven";
 
 export namespace Utils {
-    export async function getProject(pomXmlFilePath: string, iconPath?: string): Promise<ProjectItem> {
-        if (fs.existsSync(pomXmlFilePath)) {
-            const xml: string = fs.readFileSync(pomXmlFilePath, "utf8");
+    export async function getProject(absolutePath: string, workspacePath: string, iconPath?: string): Promise<ProjectItem> {
+        if (fs.existsSync(absolutePath)) {
```
Comment (jdneo) [inline src/Utils.ts]: use await

**Review by jdneo — DISMISSED** [review]

**Review by jdneo — APPROVED** [review]

File: src/ProjectDataProvider.ts [hunk 2]
```diff
@@ -25,60 +28,76 @@ export class ProjectDataProvider implements TreeDataProvider<TreeItem> {
...
-            const pinnedPomPaths: string[] = workspace.getConfiguration("maven.projects").get<string[]>("pinnedPomPaths") || [];
-            pinnedPomPaths.filter(
-                (pomPath: string) => !pomXmlFilePaths.find((value: string) => value === pomPath)
-            ).forEach((pomPath: string) => pomXmlFilePaths.push(pomPath));
-
-            pomXmlFilePaths.forEach((pomXmlFilePath: string) => {
-                todolist.push(Utils.getProject(pomXmlFilePath, this.context.asAbsolutePath(path.join("resources", "project.svg"))));
+            const ret: WorkspaceItem[] = [];
+            if (workspace.workspaceFolders) {
+                workspace.workspaceFolders.forEach((wf: WorkspaceFolder) => {
```
Comment (adashen) [inline src/ProjectDataProvider.ts:36]: can use map?

File: src/ProjectDataProvider.ts [hunk 3]
```diff
@@ -25,60 +28,76 @@ export class ProjectDataProvider implements TreeDataProvider<TreeItem> {
...
+                });
+            }
+            return ret;
+        } else if (element.contextValue === "WorkspaceItem") {
+            const workspaceItem: WorkspaceItem = <WorkspaceItem> element;
+            const promiseList: Promise<ProjectItem>[] = [];
+            const depth: number = workspace.getConfiguration("maven.projects").get<number>("maxDepthOfPom");
+            const exclusions: string[] = workspace.getConfiguration("maven.projects").get<string[]>("excludedFolders");
+            const foundPomXmls: string[] = await Utils.findAllInDir(workspaceItem.abosolutePath, "pom.xml", depth, exclusions);
+            foundPomXmls.forEach((pomXmlFilePath: string) => {
```
Comment (adashen) [inline src/ProjectDataProvider.ts:48]: same as above

File: src/Utils.ts [hunk 4]
```diff
@@ -157,21 +139,30 @@ export namespace Utils {
             });
     }
 
-    export function findAllInDir(dirname: string, targetFileName: string, depth: number): string[] {
+    export async function findAllInDir(dirname: string, targetFileName: string, depth: number, exclusion: string[] = ["**/.*"]): Promise<string[]> {
         const ret: string[] = [];
```
Comment (adashen) [inline src/Utils.ts:143]: what if the input dirname is filename?

File: src/model/FolderItem.ts [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+        pom?: IPomRoot,
+        modules?: IPomModules[]
+    };
+
+    constructor(name: string, contextValue: string, parentAbsolutePath: string, workpacePath: string, params?: object) {
+        super(name, TreeItemCollapsibleState.Collapsed);
+        this.name = name;
+        this.workspacePath = workpacePath;
+        this.parentAbsolutePath = parentAbsolutePath;
+        this.contextValue = contextValue;
```
Comment (adashen) [inline src/model/FolderItem.ts:18]: does contextValue of FolderItem fixed? if so just use a const

**Pull request merged**